engine/pkg/externalresource/s3/index.go

```diff
@@ -0,0 +1,309 @@
+// Copyright 2022 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package s3
+
+import (
+	"context"
+	"encoding/json"
+	"sync"
+	"time"
+
+	"github.com/pingcap/errors"
+	"github.com/pingcap/log"
+	brStorage "github.com/pingcap/tidb/br/pkg/storage"
+	"github.com/pingcap/tiflow/engine/pkg/externalresource/internal"
+	"github.com/pingcap/tiflow/engine/pkg/externalresource/resourcemeta/model"
+	"github.com/pingcap/tiflow/engine/pkg/tenant"
+	"go.uber.org/atomic"
+	"go.uber.org/zap"
+)
+
+const (
+	indexFileName = ".index"
+)
+
+var errIndexFileDoesNotExist = errors.New("index file does not exist")
+
+// indexManager is used to manage the set of all persisted file resources
+// on s3.
+//
+// The logic is that, when a worker or executor goes offline, as
+// a garbage collection mechanism, we remove all files on s3 that's NOT in
+// the index.
+//
+// This mechanism helps keeping the garbage collection mechanism consistent
+// with what we have for the local file resources.
+type indexManager interface {
+	SetPersisted(
+		ctx context.Context, ident internal.ResourceIdent,
+	) (bool, error)
+
+	LoadPersistedFileSet(
+		ctx context.Context, scope internal.ResourceScope,
+	) (map[model.ResourceName]struct{}, error)
+
+	Close()
+}
+
+type indexFile struct {
+	PersistedFileSet map[model.ResourceName]struct{} `json:"persisted_file_set"`
+}
+
+func newIndexFile() *indexFile {
+	return &indexFile{
+		PersistedFileSet: map[model.ResourceName]struct{}{},
+	}
+}
+
+func (f *indexFile) SetPersisted(ident internal.ResourceIdent) bool {
+	if _, ok := f.PersistedFileSet[ident.Name]; ok {
+		return false
+	}
+	f.PersistedFileSet[ident.Name] = struct{}{}
+	return true
+}
+
+func (f *indexFile) UnsetPersisted(ident internal.ResourceIdent) bool {
+	if _, ok := f.PersistedFileSet[ident.Name]; !ok {
+		return false
+	}
+	delete(f.PersistedFileSet, ident.Name)
+	return true
+}
+
+func (f *indexFile) DeserializeFrom(bytes []byte) error {
+	if err := json.Unmarshal(bytes, &f.PersistedFileSet); err != nil {
+		return errors.Annotate(err, "deserialize index file from s3")
+	}
+	return nil
+}
+
+func (f *indexFile) Serialize() ([]byte, error) {
+	bytes, err := json.Marshal(f.PersistedFileSet)
+	if err != nil {
+		return nil, errors.Annotate(err, "serialize index file to s3")
+	}
+	return bytes, nil
+}
+
+func (f *indexFile) GetPersistedFileSet() map[model.ResourceName]struct{} {
+	return f.PersistedFileSet
+}
+
+type indexManagerImpl struct {
```

Comment:
Would it better to implement `indexManager` with metastore, which may be more flexible, and don't need to worry about rate limit of S3.